platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/BirthRoundStateMigration.java

```diff
@@ -0,0 +1,115 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.state;
+
+import static com.swirlds.logging.legacy.LogMarker.STARTUP;
+
+import com.swirlds.common.merkle.crypto.MerkleCryptoFactory;
+import com.swirlds.platform.consensus.ConsensusSnapshot;
+import com.swirlds.platform.event.AncientMode;
+import com.swirlds.platform.state.signed.SignedState;
+import com.swirlds.platform.system.SoftwareVersion;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * A utility for migrating the state when birth round mode is first enabled.
+ */
+public final class BirthRoundStateMigration {
+
+    private static final Logger logger = LogManager.getLogger(BirthRoundStateMigration.class);
+
+    private BirthRoundStateMigration() {}
+
+    /**
+     * Perform required state changes for the migration to birth round mode. This method is a no-op if it is not yet
+     * time to migrate, or if the migration has already been completed.
+     *
+     * @param initialState the initial state the platform is starting with
+     * @param ancientMode  the current ancient mode
+     * @param appVersion   the current application version
+     */
+    public static void modifyStateForBirthRoundMigration(
+            @NonNull final SignedState initialState,
+            @NonNull final AncientMode ancientMode,
+            @NonNull final SoftwareVersion appVersion) {
+
+        if (ancientMode == AncientMode.GENERATION_THRESHOLD) {
+            if (initialState.getState().getPlatformState().getFirstVersionInBirthRoundMode() != null) {
+                throw new IllegalStateException(
+                        "Cannot revert to generation mode after birth round migration has been completed.");
+            }
+
+            logger.info(
+                    STARTUP.getMarker(), "Birth round state migration is not yet needed, still in generation mode.");
+            return;
+        }
+
+        final State state = initialState.getState();
+        final PlatformState platformState = state.getPlatformState();
+
+        final boolean alreadyMigrated = platformState.getFirstVersionInBirthRoundMode() != null;
+        if (alreadyMigrated) {
+            // Birth round migration was completed at a prior time, no action needed.
+            logger.info(STARTUP.getMarker(), "Birth round state migration has already been completed.");
+            return;
+        }
+
+        final long lastRoundBeforeMigration =
+                initialState.getState().getPlatformState().getRound();
```

Comment:
You're calling `initialState.getState().getPlatformState()`, but that value has been saved in local variable `platformState`

Reply:
fixed